File: collect.py
#!/usr/bin/env python3
"""
Google検索で埼玉県内の企業HPを収集し companies.csv に追記する
"""
import asyncio
import csv
import json
import re
from pathlib import Path
from urllib.parse import urlparse
from playwright.async_api import async_playwright

CONFIG_FILE = Path(__file__).parent / "config.json"
CSV_FILE = Path(__file__).parent / "companies.csv"


def load_config():
    with open(CONFIG_FILE, encoding="utf-8") as f:
        return json.load(f)


def load_existing_urls():
    if not CSV_FILE.exists():
        return set()
    with open(CSV_FILE, encoding="utf-8") as f:
        reader = csv.DictReader(f)
        return {row["url"] for row in reader if row.get("url")}


def save_company(url, company_name=""):
    with open(CSV_FILE, "a", encoding="utf-8", newline="") as f:
        writer = csv.writer(f)
        writer.writerow([url, company_name, "", "pending", "", ""])


def is_valid_company_url(url, exclude_domains):
    try:
        parsed = urlparse(url)
        domain = parsed.netloc.lower()
        # 除外ドメインチェック
        for excl in exclude_domains:
            if excl in domain:
                return False
        # .co.jp / .com / .jp などの企業ドメインのみ
        if not any(parsed.scheme.startswith(s) for s in ["http", "https"]):
            return False
        # GoogleのキャッシュURLなどを除外
        if "google" in domain or "cache" in url:
            return False
        return True
    except Exception:
        return False


def extract_base_url(url):
    """サブページをトップページに変換"""
    parsed = urlparse(url)
    return f"{parsed.scheme}://{parsed.netloc}/"


async def search_google(page, query, max_results=10, exclude_domains=None):
    """Google検索して企業URLを収集"""
    exclude_domains = exclude_domains or []
    urls = []

    search_url = f"https://www.google.com/search?q={query.replace(' ', '+')}&num={max_results}"
    print(f"  🔍 検索: {query}")

    try:
        await page.goto(search_url, timeout=15000)
        await page.wait_for_timeout(2000)

        # 検索結果のリンクを取得
        links = await page.locator("a[href]").all()
        for link in links:
            try:
                href = await link.get_attribute("href")
                if not href:
                    continue
                # Google検索結果の実URLを抽出
                if href.startswith("/url?q="):
                    href = href.split("/url?q=")[1].split("&")[0]
                if href.startswith("http") and is_valid_company_url(href, exclude_domains):
                    base = extract_base_url(href)
                    if base not in urls:
                        urls.append(base)
                        if len(urls) >= max_results:
                            break
            except Exception:
                continue

    except Exception as e:
        print(f"  ⚠️ 検索エラー: {e}")

    return urls


async def collect():
    config = load_config()
    existing_urls = load_existing_urls()
    search_cfg = config["search"]

    new_count = 0

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=False)
        page = await browser.new_page()

        # User-Agentを一般ブラウザに偽装
        await page.set_extra_http_headers({
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
        })

        for query in search_cfg["queries"]:
            urls = await search_google(
                page,
                query,
                max_results=search_cfg["results_per_query"],
                exclude_domains=search_cfg["exclude_domains"]
            )

            for url in urls:
                if url not in existing_urls:
                    save_company(url)
                    existing_urls.add(url)
                    new_count += 1
                    print(f"    ✅ 追加: {url}")
                else:
                    print(f"    ⏭️  既存: {url}")

            # 検索間の待機（Googleへの負荷軽減）
            await page.wait_for_timeout(3000)

        await browser.close()

    print(f"\n📊 収集完了: {new_count}件追加 / 合計 {len(existing_urls)}件")


if __name__ == "__main__":
    asyncio.run(collect())

File: collect.js
/**
 * Baseconnect から埼玉県内の企業HPを収集し companies.csv に追記する
 */
const fs = require('fs');
const path = require('path');
const { chromium } = require('playwright');

const CONFIG_FILE = path.join(__dirname, 'config.json');
const CSV_FILE = path.join(__dirname, 'companies.csv');

function loadConfig() {
  return JSON.parse(fs.readFileSync(CONFIG_FILE, 'utf-8'));
}

function loadExistingUrls() {
  if (!fs.existsSync(CSV_FILE)) return new Set();
  const content = fs.readFileSync(CSV_FILE, 'utf-8');
  const lines = content.trim().split('\n').slice(1); // ヘッダー除外
  return new Set(lines.map(l => l.split(',')[0]).filter(Boolean));
}

function saveCompany(url, companyName = '') {
  const line = `${url},${companyName},,,pending,,\n`;
  fs.appendFileSync(CSV_FILE, line, 'utf-8');
}

function isValidCompanyUrl(url, excludeDomains) {
  try {
    const parsed = new URL(url);
    const domain = parsed.hostname.toLowerCase();
    for (const excl of excludeDomains) {
      if (domain.includes(excl)) return false;
    }
    return url.startsWith('http');
  } catch {
    return false;
  }
}

function extractBaseUrl(url) {
  try {
    const parsed = new URL(url);
    return `${parsed.protocol}//${parsed.hostname}/`;
  } catch {
    return null;
  }
}

// Iタウンページ 埼玉県 業種別検索URL
const ITP_PAGES = [
  'https://itp.ne.jp/cgi-bin/srefer?ACT=search&prefcd=11&catcd=0401',  // 製造業
  'https://itp.ne.jp/cgi-bin/srefer?ACT=search&prefcd=11&catcd=0601',  // 建設業
  'https://itp.ne.jp/cgi-bin/srefer?ACT=search&prefcd=11&catcd=1001',  // 情報通信業
  'https://itp.ne.jp/cgi-bin/srefer?ACT=search&prefcd=11&catcd=0801',  // 不動産
  'https://itp.ne.jp/cgi-bin/srefer?ACT=search&prefcd=11&catcd=0501',  // 運輸・物流
  'https://itp.ne.jp/cgi-bin/srefer?ACT=search&prefcd=11&catcd=0701',  // 卸売・小売
  'https://itp.ne.jp/cgi-bin/srefer?ACT=search&prefcd=11&catcd=1101',  // サービス業
  'https://itp.ne.jp/cgi-bin/srefer?ACT=search&prefcd=11&catcd=0901',  // 飲食業
];

async function collectFromItp(page, listUrl, excludeDomains, existingUrls) {
  const collected = [];
  console.log(`  📂 ${listUrl}`);

  // 最大5ページ分巡回
  for (let p = 1; p <= 5; p++) {
    const url = p === 1 ? listUrl : `${listUrl}&page=${p}`;
    try {
      await page.goto(url, { timeout: 20000, waitUntil: 'domcontentloaded' });
      await page.waitForTimeout(1500);

      // 企業名とHP URLを取得
      const entries = await page.evaluate(() => {
        const results = [];
        // 各企業ブロックを探す
        const items = document.querySelectorAll('.searchResult__item, .result-item, li[class*="result"]');
        items.forEach(item => {
          const nameEl = item.querySelector('h2, h3, .name, .company-name, strong');
          const name = nameEl ? nameEl.textContent.trim() : '';
          // HP URLを探す
          const links = Array.from(item.querySelectorAll('a[href]'));
          for (const a of links) {
            const href = a.href;
            if (
              href.startsWith('http') &&
              !href.includes('itp.ne.jp') &&
              !href.includes('google.com') &&
              !href.includes('facebook.com') &&
              !href.includes('twitter.com')
            ) {
              results.push({ url: href, name });
              break;
            }
          }
        });
        return results;
      });

      if (entries.length === 0) break; // これ以上ページなし

      for (const { url: hpUrl, name } of entries) {
        const base = extractBaseUrl(hpUrl);
        if (base && !existingUrls.has(base) && isValidCompanyUrl(base, excludeDomains)) {
          collected.push({ url: base, name });
        }
      }

      console.log(`    📄 ${p}ページ目: ${entries.length}件取得`);
    } catch (e) {
      console.log(`  ⚠️ エラー(p${p}): ${e.message.slice(0, 60)}`);
      break;
    }
  }

  return collected;
}

async function collect() {
  const config = loadConfig();
  const existingUrls = loadExistingUrls();
  const { exclude_domains } = config.search;

  // 既存データをリセット（musubu.inのゴミを削除）
  fs.writeFileSync(CSV_FILE, 'url,company_name,contact_url,status,submitted_at,note\n', 'utf-8');
  existingUrls.clear();

  let newCount = 0;

  const browser = await chromium.launch({ headless: false });
  const context = await browser.newContext({
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
  });
  const page = await context.newPage();

  for (const listUrl of ITP_PAGES) {
    const results = await collectFromItp(page, listUrl, exclude_domains, existingUrls);

    for (const { url, name } of results) {
      saveCompany(url, name);
      existingUrls.add(url);
      newCount++;
      console.log(`    ✅ 追加: ${url} (${name})`);
    }

    await page.waitForTimeout(2000);
  }

  await browser.close();
  console.log(`\n📊 収集完了: ${newCount}件追加 / 合計 ${existingUrls.size}件`);
}

collect().catch(console.error);
